### ISSUE TITLE
iOS crashes when trying to initiate another call (after hanging up on the first one)

### ISSUE THREAD

**Issue by tompi** [Azure/Communication#61]
Maybe me not cleaning up something, but the SDK throws an exception on initiating call nr 2:

````
Objective-C exception thrown.  Name: NSInternalInconsistencyException Reason: SKPPreviewVideoRenderer: Failed to create vb::IAppleRendererBinding
Native stack trace:
	0   CoreFoundation                      0x00000001aef71358 C89C0E79-E3BA-397B-84AB-0B884D980499 + 1250136
	1   libobjc.A.dylib                     0x00000001aec86cc0 objc_exception_throw + 60
	2   CoreFoundation                      0x00000001aee6e878 C89C0E79-E3BA-397B-84AB-0B884D980499 + 190584
	3   Foundation                          0x00000001af311c74 2F43F187-0E88-35B3-82EE-D53E3FCEF0D2 + 904308
	4   AzureCommunicationCalling           0x0000000105766bd4 _Z27skp_NSStringFromOpenGLErrorj + 36552
	5   AzureCommunicationCalling           0x0000000105a47538 _ZN8MySkyLib20OnMediaStatusChangedEN6SkyLib11MEDIASTATUSE + 2949932
	6   AzureCommunicationCalling           0x0000000105a4b090 _ZN8MySkyLib20OnMediaStatusChangedEN6SkyLib11MEDIASTATUSE + 2965124
	7   AzureCommunicationCalling           0x0000000105a24570 _ZN8MySkyLib20OnMediaStatusChangedEN6SkyLib11MEDIASTATUSE + 2806628
	8   AzureCommunicationCalling           0x0000000105d57c10 _ZN8MySkyLib20OnMediaStatusChangedEN6SkyLib11MEDIASTATUSE + 6162948
	9   AzureCommunicationCalling           0x0000000105db4aa4 _ZN8MySkyLib20OnMediaStatusChangedEN6SkyLib11MEDIASTATUSE + 6543512
	10  AzureCommunicationCalling           0x0000000105db4920 _ZN8MySkyLib20OnMediaStatusChangedEN6SkyLib11MEDIASTATUSE + 6543124
	11  AzureCommunicationCalling           0x0000000105db4d74 _ZN8MySkyLib20OnMediaStatusChangedEN6SkyLib11MEDIASTATUSE + 6544232
	12  AzureCommunicationCalling           0x0000000105db3afc _ZN8MySkyLib20OnMediaStatusChangedEN6SkyLib11MEDIASTATUSE + 6539504
	13  AzureCommunicationCalling           0x0000000105db3c64 _ZN8MySkyLib20OnMediaStatusChangedEN6SkyLib11MEDIASTATUSE + 6539864
	14  libdispatch.dylib                   0x00000001aec12134 6D9E71A9-98B8-3473-A940-CA049AE82C79 + 8500
	15  libdispatch.dylib                   0x00000001aec135ac 6D9E71A9-98B8-3473-A940-CA049AE82C79 + 13740
	16  libdispatch.dylib                   0x00000001aec15a1c 6D9E71A9-98B8-3473-A940-CA049AE82C79 + 23068
	17  libdispatch.dylib                   0x00000001aec22724 6D9E71A9-98B8-3473-A940-CA049AE82C79 + 75556
	18  libdispatch.dylib                   0x00000001aec22ed0 6D9E71A9-98B8-3473-A940-CA049AE82C79 + 77520
	19  libsystem_pthread.dylib             0x00000001aec796dc _pthread_wqthread + 216
	20  libsystem_pthread.dylib             0x00000001aec7f9c8 start_wqthread + 8
````

**Comment by raosanat** [comment]
Hi @tompi Can you give more information, like
1. How often is this reproducible ?
2. Exact steps, API calls done to repro the issue
3. Can you attach the logs ? 

Steps to get logs
1. In Xcode, go to Windows -> Devices and Simulators -> Devices -> Select your device -> Under Installed Apps -> Select your application -> Select  at the bottom -> Download container
2. You will get a file with extension xcappdata, Right-Click on Settings icon and select “Show package contents” 
3. Go to Appdata->Library->Caches and upload the .blog files

**Comment by tompi** [comment]
@raosanat This crash was probably due to not disposing the renderers and/or the videostream after a call hangup. Works fine after I started doing that.

**Comment by tompi** [comment]
Sorry, seems this is still happening.

I spent some time debugging, and it seems to happen every time I try to make an ACSRendererView from the local video stream(works fine first time, but not the second time). Even though I dispose this, and most of the others, I only reuse ACSDeviceManager, ACSCallClient and ACSCallClient, that should be fine right?

Anyway, exception is still the same, but I attached it and also the complete logs.
I suspect this is your code that is throwing, but I can figure out why...:

````
*** Assertion failure in virtual void PreviewVideoRenderedBindingEvent::OnBindingFailed()(), /Volumes/ServerHD2/agent/_work/1/s/SpoolCallingStack/spool/sdk/iOS/ACSCallingSDK/SKPPullVideoRenderer/SKPPreviewVideoRenderer.mm:95
````
[log.txt](https://github.com/Azure/Communication/files/5377309/log.txt)
[stacktrace.txt](https://github.com/Azure/Communication/files/5377310/stacktrace.txt)

**Comment by sankum-msft** [comment]
@tompi : This error indicates that previous `ACSRenderer` object for that `LocalVideoStream`/`VideoDeviceInfo` is not yet disposed. Your message also backs up this hypothesis. Please try disposing `ACSRenderer`. You don't need to dispose the view; disposing renderer disposes view as well.

But, agreed that this should not be a crash. I have created task to handle this gracefully and raise a more appropriate error.

**Comment by tompi** [comment]
@sankum-msft Hey, I did some more debugging on this, and I cant make it work. I tried disposing ACSRenderer, but same problem. I then tried disposing every object: ACSRenderer, ACSLocalVideoStream, ACSCall, ACSCallAgent, ACSDeviceManager and the ACSCallClient. 

I could NOT manage to dispose the ACSCallClient: the app just hangs(indefinitely...). SO my theory is that the ACSCallClient is hanging on to some resource which makes the second group-call crash. If you have tested this on your end, Im probably doing something wrong(but I really have a minimal setup now... Is there a specific sequence I need to dispose all the objects in to make this work? (If you have some example code in which multiple sequential calls work that would be great!)

**Comment by raosanat** [comment]
Hi @tompi if you download your app data there will be a file called "acs_sdk.blog" in Caches folder, can you attach that ?

**Comment by tompi** [comment]
Hey @raosanat , sorry about the delay, I downloaded the container, zipped it and attached to message:
[acs-container.zip](https://github.com/Azure/Communication/files/5421022/acs-container.zip)

**Comment by tompi** [comment]
Just to be clear: this happens when I make two consecutive calls to the same group.
I dispose everything, except the ACSCallAgent, since trying to dispose this object just hangs the app(no crash, just hang)

**Comment by tompi** [comment]
@raosanat and @sankum-msft : I just figured out that this was entirely my own fault... Sorry!
I had commented out the dispose method in the bindings(due to a warning). Once I commented in this I can du multiple calls fine after disposing the renderer.

My fix is here if you are curious: https://github.com/Laerdal/Xamarin.AzureCommunicationCalling/commit/d68a94ffe916c25209c7efebb416878c25b642be

Again, sorry about wasting your time :(
